File: server/internal/model/mission.go
package model

import (
	"encoding/json"
	"time"
)

// MissionTier 代表派遣任务时选择的"档位"，对应不同的 DeepSeek 模型与思考强度。
type MissionTier string

const (
	TierFlash    MissionTier = "flash"    // deepseek-v4-flash + thinking disabled
	TierStandard MissionTier = "standard" // deepseek-v4-flash + thinking high  (默认)
	TierPro      MissionTier = "pro"      // deepseek-v4-pro   + thinking max
)

// MissionStatus 是任务的生命周期状态。
type MissionStatus string

const (
	StatusPending MissionStatus = "pending"
	StatusRunning MissionStatus = "running"
	StatusDone    MissionStatus = "done"
	StatusAborted MissionStatus = "aborted"
	StatusError   MissionStatus = "error"
)

// Mission 是一次完整的任务派遣（特工出勤）。
type Mission struct {
	ID           string        `json:"id"`
	UserID       int64         `json:"user_id"`
	Codename     string        `json:"codename"`
	Brief        string        `json:"brief"`
	Tier         MissionTier   `json:"tier"`
	Status       MissionStatus `json:"status"`
	Loadout      []string      `json:"loadout"`       // 允许动用的装备名列表
	WorkspaceDir string        `json:"workspace_dir"` // 任务隔离目录绝对路径（不直接暴露给前端）
	InputTokens  int64         `json:"input_tokens"`
	OutputTokens int64         `json:"output_tokens"`
	// SeriesID 行动卷宗 ID：一次性派遣的 mission，SeriesID = ID；
	// 通过「继续安排」生成的后续 mission，SeriesID = 首次 mission 的 ID。
	SeriesID  string  `json:"series_id"`
	SeriesSeq int     `json:"series_seq"`           // 在该卷宗内的序号，1 = 首次
	ParentID  *string `json:"parent_id,omitempty"`  // 上一个 mission 的 ID（首次为 nil）
	StartedAt *time.Time `json:"started_at,omitempty"`
	EndedAt   *time.Time `json:"ended_at,omitempty"`
	CreatedAt time.Time  `json:"created_at"`
}

// StepType 是行动现场事件流上每一条记录的类型。
//
// 它与 Step.Payload 的具体结构对应，约定如下：
//
//	thought       { "text": string }              ← 模型 reasoning_content 增量段
//	message       { "text": string }              ← 模型 content 输出段（可能多段拼装）
//	tool_call     { "id","name","arguments_json" }← 模型决定调装备
//	tool_result   { "id","name","ok","content","error" } ← 装备执行结果
//	artifact      { "artifact_id","name","kind" } ← 入柜事件
//	system        { "kind","text" }               ← 派遣/撤离/错误提示
//	usage         { "input_tokens","output_tokens","cache_hit_tokens" } ← 计费快照
type StepType string

const (
	StepThought    StepType = "thought"
	StepMessage    StepType = "message"
	StepToolCall   StepType = "tool_call"
	StepToolResult StepType = "tool_result"
	StepArtifact   StepType = "artifact"
	StepSystem     StepType = "system"
	StepUsage      StepType = "usage"
)

// Step 是任务事件流的一条不可变记录。所有事件都按 (mission_id, seq) 单调递增。
type Step struct {
	ID        int64           `json:"id"`
	MissionID string          `json:"mission_id"`
	Seq       int             `json:"seq"`
	Ts        time.Time       `json:"ts"`
	Type      StepType        `json:"type"`
	Payload   json.RawMessage `json:"payload"`

	// ReasoningContent 仅对"导致 tool_call 的那一轮 assistant 消息"非空。
	// DeepSeek V4 thinking 模式要求：tool_call 所在轮的 reasoning_content
	// 必须在下一次请求中完整回传，否则 HTTP 400。
	// 我们把它和发起 tool_call 的 step 一起持久化，以便重连/重放/继续。
	ReasoningContent string `json:"-"`
}

// Artifact 是任务产出的工件（入柜物）。
type Artifact struct {
	ID        int64     `json:"id"`
	MissionID string    `json:"mission_id"`
	Kind      string    `json:"kind"` // file / image / code / url / chart
	Name      string    `json:"name"`
	Path      string    `json:"path"` // 相对 mission workspace 的路径
	Mime      string    `json:"mime,omitempty"`
	Size      int64     `json:"size"`
	CreatedAt time.Time `json:"created_at"`
}

// Review 是用户对一次行动的点评：1-5 星 + 可选评语。
// 一个 mission 最多一条 review，重复提交即 upsert。
type Review struct {
	MissionID string    `json:"mission_id"`
	Rating    int       `json:"rating"` // 1..5
	Comment   string    `json:"comment"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// Skill 是用户从一次高分行动里"沉淀"下来的可复用方法论。
// 之后派遣相关任务时，prompt_template 可以作为隐式系统提示参与。
type Skill struct {
	ID              int64     `json:"id"`
	UserID          int64     `json:"user_id"`
	Name            string    `json:"name"`
	Description     string    `json:"description"`
	TriggerHint     string    `json:"trigger_hint"`    // 什么时候启用这项技能的一句话提示
	PromptTemplate  string    `json:"prompt_template"` // 实际注入的 prompt 片段
	SourceMissionID *string   `json:"source_mission_id,omitempty"`
	CreatedAt       time.Time `json:"created_at"`
}

File: app/lib/models/mission.dart
// 与后端 internal/model/mission.go 的 JSON 形态一一对应。
// 任何字段含义/契约变更都需要后端同步。

enum MissionTier { flash, standard, pro }

extension MissionTierX on MissionTier {
  String get wire {
    switch (this) {
      case MissionTier.flash:
        return 'flash';
      case MissionTier.standard:
        return 'standard';
      case MissionTier.pro:
        return 'pro';
    }
  }

  String get label {
    switch (this) {
      case MissionTier.flash:
        return '闪电';
      case MissionTier.standard:
        return '标准';
      case MissionTier.pro:
        return '深度';
    }
  }

  String get desc {
    switch (this) {
      case MissionTier.flash:
        return '不思考 · 最快响应';
      case MissionTier.standard:
        return '默认思考 · 性价比最高';
      case MissionTier.pro:
        return '深度推理 · 适合复杂任务';
    }
  }

  static MissionTier fromWire(String s) =>
      MissionTier.values.firstWhere((t) => t.wire == s, orElse: () => MissionTier.standard);
}

enum MissionStatus { pending, running, done, aborted, error }

extension MissionStatusX on MissionStatus {
  String get wire => name;

  String get label {
    switch (this) {
      case MissionStatus.pending:
        return '待命';
      case MissionStatus.running:
        return '行动中';
      case MissionStatus.done:
        return '已完成';
      case MissionStatus.aborted:
        return '已撤离';
      case MissionStatus.error:
        return '失联';
    }
  }

  static MissionStatus fromWire(String s) =>
      MissionStatus.values.firstWhere((t) => t.name == s, orElse: () => MissionStatus.pending);

  bool get isTerminal => this == MissionStatus.done || this == MissionStatus.aborted || this == MissionStatus.error;
}

class Mission {
  Mission({
    required this.id,
    required this.codename,
    required this.brief,
    required this.tier,
    required this.status,
    required this.loadout,
    required this.inputTokens,
    required this.outputTokens,
    required this.createdAt,
    required this.seriesId,
    required this.seriesSeq,
    this.parentId,
    this.startedAt,
    this.endedAt,
  });

  final String id;
  final String codename;
  final String brief;
  final MissionTier tier;
  final MissionStatus status;
  final List<String> loadout;
  final int inputTokens;
  final int outputTokens;
  final DateTime createdAt;
  final String seriesId;
  final int seriesSeq;
  final String? parentId;
  final DateTime? startedAt;
  final DateTime? endedAt;

  factory Mission.fromJson(Map<String, dynamic> json) => Mission(
        id: json['id'] as String,
        codename: (json['codename'] as String?) ?? '未命名行动',
        brief: (json['brief'] as String?) ?? '',
        tier: MissionTierX.fromWire((json['tier'] as String?) ?? 'standard'),
        status: MissionStatusX.fromWire((json['status'] as String?) ?? 'pending'),
        loadout: ((json['loadout'] as List?) ?? const []).cast<String>(),
        inputTokens: (json['input_tokens'] as num?)?.toInt() ?? 0,
        outputTokens: (json['output_tokens'] as num?)?.toInt() ?? 0,
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
        seriesId: (json['series_id'] as String?) ?? (json['id'] as String),
        seriesSeq: (json['series_seq'] as num?)?.toInt() ?? 1,
        parentId: json['parent_id'] as String?,
        startedAt: (json['started_at'] as String?) != null
            ? DateTime.parse(json['started_at'] as String).toLocal()
            : null,
        endedAt: (json['ended_at'] as String?) != null
            ? DateTime.parse(json['ended_at'] as String).toLocal()
            : null,
      );
}

/// 用户对一次行动的点评（1-5 星 + 评语）。
class MissionReview {
  MissionReview({
    required this.missionId,
    required this.rating,
    required this.comment,
    required this.createdAt,
    required this.updatedAt,
  });

  final String missionId;
  final int rating;
  final String comment;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory MissionReview.fromJson(Map<String, dynamic> json) => MissionReview(
        missionId: json['mission_id'] as String,
        rating: (json['rating'] as num).toInt(),
        comment: (json['comment'] as String?) ?? '',
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
        updatedAt: DateTime.parse(json['updated_at'] as String).toLocal(),
      );
}

/// 沉淀下来的技能（高分行动总结成的 Skill）。
class Skill {
  Skill({
    required this.id,
    required this.name,
    required this.description,
    required this.triggerHint,
    required this.promptTemplate,
    required this.createdAt,
    this.sourceMissionId,
  });

  final int id;
  final String name;
  final String description;
  final String triggerHint;
  final String promptTemplate;
  final String? sourceMissionId;
  final DateTime createdAt;

  factory Skill.fromJson(Map<String, dynamic> json) => Skill(
        id: (json['id'] as num).toInt(),
        name: (json['name'] as String?) ?? '',
        description: (json['description'] as String?) ?? '',
        triggerHint: (json['trigger_hint'] as String?) ?? '',
        promptTemplate: (json['prompt_template'] as String?) ?? '',
        sourceMissionId: json['source_mission_id'] as String?,
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
      );
}

/// Step 是事件流上的单条原子事件。payload 解码时按 type 自行 cast。
class MissionStep {
  MissionStep({
    required this.id,
    required this.missionId,
    required this.seq,
    required this.ts,
    required this.type,
    required this.payload,
  });

  final int id;
  final String missionId;
  final int seq;
  final DateTime ts;
  final String type;
  final Map<String, dynamic> payload;

  factory MissionStep.fromJson(Map<String, dynamic> json) => MissionStep(
        id: (json['id'] as num).toInt(),
        missionId: json['mission_id'] as String,
        seq: (json['seq'] as num).toInt(),
        ts: DateTime.parse(json['ts'] as String).toLocal(),
        type: json['type'] as String,
        payload: (json['payload'] as Map?)?.cast<String, dynamic>() ?? const {},
      );
}

class Artifact {
  Artifact({
    required this.id,
    required this.missionId,
    required this.kind,
    required this.name,
    required this.path,
    required this.mime,
    required this.size,
    required this.createdAt,
  });

  final int id;
  final String missionId;
  final String kind;
  final String name;
  final String path;
  final String mime;
  final int size;
  final DateTime createdAt;

  factory Artifact.fromJson(Map<String, dynamic> json) => Artifact(
        id: (json['id'] as num).toInt(),
        missionId: json['mission_id'] as String,
        kind: (json['kind'] as String?) ?? 'file',
        name: (json['name'] as String?) ?? '',
        path: (json['path'] as String?) ?? '',
        mime: (json['mime'] as String?) ?? '',
        size: (json['size'] as num?)?.toInt() ?? 0,
        createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
      );
}

class ToolInfo {
  ToolInfo({required this.name, required this.displayName, required this.description});
  final String name;
  final String displayName;
  final String description;

  factory ToolInfo.fromJson(Map<String, dynamic> json) => ToolInfo(
        name: json['name'] as String,
        displayName: (json['display_name'] as String?) ?? json['name'] as String,
        description: (json['description'] as String?) ?? '',
      );
}
